File: Platformer/scripts/Fish.jl
module FishModule
    using JulGame

    mutable struct Fish <: Script
        animator
        endingY::EditorExport{Int}
        isFire::EditorExport{Bool}
        isMovingUp::Bool
        parent::JulGame.EntityModule.Entity
        sound::SoundSourceModule.SoundSource
        speed::EditorExport{Float64}
        startingY::EditorExport{Int}

        function Fish()
            this = new()

            this.endingY = 0
            this.isFire = false
            this.isMovingUp = false
            this.speed = 0.0
            this.startingY = 0

            return this
        end
    end

    function JulGame.initialize(this::Fish)
        this.animator = this.parent.animator
        this.parent.sprite.rotation = 90.0
        this.parent.transform.position = Vector2f(this.parent.transform.position.x, this.startingY)
    end

    function JulGame.update(this::Fish, deltaTime)
        if this.parent.transform.position.y >= this.startingY && !this.isMovingUp
            this.parent.sprite.rotation = this.isFire ? 0 : 90.0
            this.isMovingUp = true
        elseif this.parent.transform.position.y <= this.endingY && this.isMovingUp
            this.parent.sprite.rotation = this.isFire ? 180.0 : 270.0
            this.isMovingUp = false
        end

        if this.isMovingUp
            this.parent.transform.position = Vector2f(this.parent.transform.position.x, this.parent.transform.position.y - this.speed*deltaTime)
        else
            this.parent.transform.position = Vector2f(this.parent.transform.position.x, this.parent.transform.position.y + this.speed*deltaTime)
        end
    end
end # module

File: Platformer/ts/_generated/scripts/Fish.ts
import { registerScript } from "julgame/src/engine/runtime/scriptRegistry";


    // using JulGame

    export class Fish {
        animator
        endingY: number
        isFire: boolean
        isMovingUp: boolean
        parent: Entity
        sound: SoundSource
        speed: number
        startingY: number

        constructor() {
            

            this.endingY = 0
            this.isFire = false
            this.isMovingUp = false
            this.speed = 0.0
            this.startingY = 0

        }
    }

    function JulGame_initialize_Fish(self: Fish) {
        self.animator = self.parent.animator
        self.parent.sprite.rotation = 90.0
        self.parent.transform.position = {x: self.parent.transform.position.x, y: self.startingY}
    }

    function JulGame_update_Fish(self: Fish, deltaTime) {
        if (self.parent.transform.position.y >= self.startingY && !self.isMovingUp) {
            self.parent.sprite.rotation = self.isFire ? 0 : 90.0
            self.isMovingUp = true
        } else if (self.parent.transform.position.y <= self.endingY && self.isMovingUp) {
            self.parent.sprite.rotation = self.isFire ? 180.0 : 270.0
            self.isMovingUp = false
        }

        if (self.isMovingUp) {
            self.parent.transform.position = {x: self.parent.transform.position.x, y: self.parent.transform.position.y - self.speed*deltaTime}
        } else {
            self.parent.transform.position = {x: self.parent.transform.position.x, y: self.parent.transform.position.y + self.speed*deltaTime}
        }
    }
registerScript("Fish", {
    create: () => new Fish(),
    initialize: JulGame_initialize_Fish,
    update: JulGame_update_Fish,
});
